### PR TITLE
Uses import tool counts computer for rebuilding missing counts store

### PR DESCRIPTION
**Pull request by tinwelint**
This is the main purpose of this commit. However a couple of related
things have also been made in here to improve performance around counts
rebuilding and convenience in the code:

- Parallelized some aspects of counts computer from the import tool.
  Basically the node label are counted and efficiently cached, so that
  the relationship counts can be processed sequentially, using the
  node-label cache. Relationship counting is parallelized into
  reading(single) and counting(multi). Each thread keeping local counts,
  aggregated in the end.
- ProducerStep has been simplified. Less code, more generic and easier to
  extend.
- Convenience around executing and supervising stages by the introduced
  ExecutorSupervisors.

Local measurements show 4x improvement from before and scalability should
be linear. The limitation is the size of the node cache, which "should" be
OK given that you generally run larger stores on larger machines. Looking
ahead there will be multi-pass support, where parts of the store is
processed in each pass, will be added to the staging framework to remove
problems for huge stores on not-that-huge machines memory-wise.

### CODE REVIEW

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreFactory.java [hunk 1]
```diff
@@ -621,6 +622,18 @@ private void createEmptyStore( File fileName, String typeAndVersionDescriptor, B
         }
     }
 
+    /**
+     * I.e. total number of used/unused records + 1
+     * @param storeFile {@link StoreFile} to get the name from.
+     * @param recordSize record size of that store.
+     * @return highId, i.e. an id one greater than the highest id in the store.
+     */
+    public long getHighId( StoreFile storeFile, int recordSize )
```
Comment (lutovich) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreFactory.java]: Maybe it is worth making a comment/assertion that this method is only valid for opened store files that have no trailers?

**Comment by lutovich** [comment]
Looks good to me

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreFactory.java (hunk 1 above)
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreFactory.java]: Oh yeah, hmm you know what? We could instead just read the highId from `.id` file. I think that should work just fine instead.

**Comment by davidegrohmann** [comment]
@tinwelint it looks good to me as well, just wondering if we get any benefit in adding some unit tests to the new stuff, in particular `ProcessRelationshipCountsDataStep`, `ReadRelationshipCountsDataStep` and `ExecutionSupervisors`

**Comment by tinwelint** [comment]
@davidegrohmann Personally I don't feel the need to add such tests. As with most things in the staging/steps code of the importer, every piece of code on its own is trivial; it's the interaction between everything that is the complicated part and for that we have tests. It might just be me.

**Comment by davidegrohmann** [comment]
@tinwelint right. I can buy that. Let's merge this.

**Pull request merged**